rcl/src/rcl/remap.c

```
@@ -0,0 +1,249 @@
+#include "rcl/expand_topic_name.h"
+#include "rcutils/strdup.h"
+#include "rcutils/types/string_map.h"
+#include "./arguments_impl.h"
+#include "./remap_impl.h"
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+rcl_remap_t
+rcl_remap_get_zero_initialized()
+{
+  rcl_remap_t rule;
+  rule.type = RCL_UNKNOWN_REMAP;
+  rule.node_name = NULL;
+  rule.match = NULL;
+  rule.replacement = NULL;
+  return rule;
+}
+
+rcl_ret_t
+rcl_remap_fini(
+  rcl_remap_t * rule,
+  rcl_allocator_t allocator)
+{
+  if (NULL != rule->node_name) {
+    allocator.deallocate(rule->node_name, allocator.state);
+    rule->node_name = NULL;
+  }
+  if (NULL != rule->match) {
+    allocator.deallocate(rule->match, allocator.state);
+    rule->match = NULL;
+  }
+  if (NULL != rule->replacement) {
+    allocator.deallocate(rule->replacement, allocator.state);
+    rule->replacement = NULL;
+  }
+  return RCL_RET_OK;
+}
+
+/// Get the first matching rule in a chain
+/// \return NULL if no rule in the chain matched
+RCL_LOCAL
+rcl_remap_t *
+_rcl_remap_first_match(
+  rcl_remap_t * remap_rules,
+  int num_rules,
+  rcl_remap_type_t type_bitmask,
+  const char * name,
+  const char * node_name,
+  const char * node_namespace,
+  const rcutils_string_map_t * substitutions,
+  rcutils_allocator_t allocator)
+{
+  for (int i = 0; i < num_rules; ++i) {
+    rcl_remap_t * rule = &(remap_rules[i]);
+    if (rule->type & type_bitmask) {
+      if (NULL == rule->node_name || NULL == node_name || 0 == strcmp(rule->node_name, node_name)) {
+        bool matched = false;
+        if (rule->type & (RCL_TOPIC_REMAP | RCL_SERVICE_REMAP)) {
+          // topic and service rules need the match side to be expanded to a FQN
+          char * expanded_match = NULL;
+          rcl_ret_t ret = rcl_expand_topic_name(
+            rule->match, node_name, node_namespace, substitutions, allocator, &expanded_match);
+          if (ret != RCL_RET_OK) {
+            // Expansion failed with this rule, try another
```

It would be worth considering why this failed, for example if it failed to allocate memory you might want to just stop rather than continue trying. On the other hand, if the match is just not a valid topic it might be safe to ignore and continue.

---

At the very least you might need to do a `rcl_reset_error()` in cases where `rcl_expand_topic_name()` set the error state.

---

Better error handling in 8cbe7f80e55451ec07a307d56af15e952ae95d8a

Function returns right away if the node name or namespace is invalid, and if BAD_ALLOC comes up. That error code + message is given back to the caller of the remapping function called.